torch/quantization/fx/quantize.py

```diff
@@ -407,6 +414,696 @@ def load_arg(a):
+    node: Node,
+    target_dtype: torch.dtype,
+    node_name_to_target_dtype: Dict[str, torch.dtype],
+    matches: Dict[str, MatchResult],
+) -> None:
+    """
+    Assigns `target_dtype` to `node`. If `node` is matched to an instance
+    of `CopyNodeQuantizeHandler`, also call this function recursively on
+    the first argument, to propagate the dtype to the caller.
+    """
+    node_name_to_target_dtype[node.name] = target_dtype
+    # if this is a copy node, propagate to first arg
+    root_node, matched_nodes, pattern, qhandler, qconfig = matches.get(
+        node.name, (None, None, None, None, None))
+    if isinstance(qhandler, CopyNodeQuantizeHandler):
+        prev_node = node.args[0]
+        if isinstance(prev_node, Node):
+            maybe_propagate_dtype_for_node_rewrite(
+                prev_node, target_dtype, node_name_to_target_dtype, matches)
+
+def propagate_dtypes_for_known_nodes_rewrite(
+    graph: Graph,
+    node_name_to_target_dtype: Dict[str, torch.dtype],
+    matches: Dict[str, MatchResult],
+) -> None:
+    """
+    Currently we assume that inputs to the graph are either `torch.float` or
+    `torch.quint8`, which is not always correct. For ops such as
+    `x.masked_fill(mask, value)`, we know that the dtype of  `mask` is a
+    `BoolTensor`. Propagate this information throughout the graph.
+
+    Note: not all dtypes in the graph will be correct after this pass, but a
+    higher percentage of them will be correct. Hopefully in the future we can
+    replace this with a better way to reason about dtypes of tensors.
+    """
+    for node in graph.nodes:
+        bool_arg_idxs = node_bool_tensor_arg_indexes(node)
+        for bool_arg_idx in bool_arg_idxs:
+            cur_node = node.args[bool_arg_idx]
+            maybe_propagate_dtype_for_node_rewrite(
+                cur_node, torch.bool, node_name_to_target_dtype, matches)
+
+def adjust_observers_for_cat_rewrite(
+    node: Node,
+    model: torch.nn.Module,
+    modules: Dict[str, torch.nn.Module],
+) -> None:
+    """
+    Ensures that for quantized `torch.cat` nodes, we share an observer
+    for all input arguments as well as the output argument. In detail, given
+    a graph of
+
+      x0 -> obs0 -> cat -> x2
+                  /
+      x1 -> obs1 /
+
+    where node obs0 points to observer instance observer0,
+    obs1 points to observer1 and obs2 points to observer2, we make nodes obs1
+    and ob2 point to observer0.
+    """
+    # find the observer module to use
+    first_arg = node.args[0]
+    assert isinstance(first_arg, (list, tuple))
+    first_arg_arg = first_arg[0]
+
+    # if we have a graph such as
+    #   observed_node -> non_observed_node -> cat
+    # we need to navigate up to the first observer
+    iteration_guard = 0
+    while not is_activation_post_process_node(first_arg_arg, modules):
+        first_arg_arg = first_arg_arg.args[0]
+        iteration_guard += 1
+        if iteration_guard > 10:
```

Comment:
What is the reason to stop at 10 here?

Reply:
Will this fail in case we have > 10 non_observed nodes between the first observer and current op? Wondering if that is a valid case we expect to see in models.

Reply:
sure, we can increase the limit to 10e5